### ISSUE TITLE
Carlini Wagner L2 Attack doesn't distort images.

### ISSUE THREAD

**Issue by blgnksy** [tensorflow/cleverhans#881]
Hi,
I am trying to craft adversarial examples using the CarliniWagnerL2 method. My input images are in [-1,1] . When I check the outputs, there is no distortion, in fact Euclidean distance is 0. The selected attack parameters are:

```
    cw_params = {'binary_search_steps': 3, 
                 'abort_early' : True, 
                 'max_iterations': 250, 
                 'learning_rate': 0.001, 
                 'batch_size': batch_size, 
                 'initial_const': 100, 
                 'nb_classes': num_classes, 
                 'confidence': 0, 
                 'clip_min': -1.0, 
                 'clip_max': 1.0 
                 }
``` 
I read the parameters' documentation but I couldn't figured out the problem.

**Comment by goodfeli** [comment]
Can you paste a complete script that we can run? We have unit tests that make sure CarliniWagnerL2 successfully modifies the images in some cases so this must be something specific to your model, etc.

**Comment by blgnksy** [comment]
Hi,

This is my attack class:
```
class InceptionModel(object):
    def __init__(self, num_classes):
        self.num_classes = num_classes
        self.built = False

    def __call__(self, x_input):
        """Constructs model and return probabilities for given input."""
        reuse = True if self.built else None
        with slim.arg_scope(inception.inception_v3_arg_scope()):
            _, end_points = inception.inception_v3(
                            x_input, num_classes=self.num_classes, is_training=False,
                            reuse=reuse)
        self.built = True
        output = end_points['Predictions']
        probs = output.op.inputs[0]
        return probs
```
And this is the attack:
```
with tf.Session() as sess:
    filenames, images, labels= next(image_iterator)
    print(labels)
    x_input = tf.placeholder(tf.float32, shape=batch_shape)

    cw_params = {'binary_search_steps': 3, 
                 'abort_early' : True, 
                 'max_iterations': 250, 
                 'learning_rate': 0.001, 
                 'batch_size': batch_size, 
                 'initial_const': 100, 
                 'nb_classes': num_classes, 
                 'confidence': 0., 
                 'clip_min': -1.0, 
                 'clip_max': 1.0 
                 } 

    model = InceptionModel(num_classes)
    clw2=CarliniWagnerL2(model) 
    x_adv = clw2.generate(x_input)

    sess.run(tf.global_variables_initializer())

    saver = tf.train.Saver(slim.get_model_variables())

    nontargeted_images = sess.run(x_adv, feed_dict={x_input: images})
```
cleverhans version is 2.0.0-dd5592125fb46237090c4d6ecd9393aa
python version is 2.7.12
tensorflow version is 1.4.0

**Comment by goodfeli** [comment]
Could you post a script that we could actually run, rather than isolated code snippets? It's not too hard for me to debug something that runs, but from this starting point I'd spend more time setting up all the imports etc than actually debugging.

I don't see anything conspicuously wrong in these code snippets.

Three thoughts:
1) It's better to work with logits than to work with probabilities. Your model is returning only probabilities, which is probably not good for numerical stability.
2) Using a `callable` is deprecated. This shouldn't be the problem, but there's a small chance there's a bug in the handling of `callable` objects. It's thus possible you could fix the problem by changing `InceptionModel` to inherit from `cleverhans.model.Model` and implementing `get_logits` rather than `__call__`.
3) cleverhans doesn't really officially support tf 1.4 anymore. The tests only run on tf 1.8 and later. There's a small chance that updating tensorflow will fix the issue.

**Comment by B-Yassine** [comment]
Have you tried to classify the output image and you don't have a misclassification? Or you just wanna visualize the perturbations and you got an imperceptible perturbed output?

**Comment by blgnksy** [comment]
@B-Yassine I tried to classifying the adversarial image. But The defender classified the adversarial to the correct label. Then I tried to visualize the perturbation but I realised that there were no perturbation.

**Comment by goodfeli** [comment]
Can you please post a short, executable script that causes the problem on your machine?

**Comment by blgnksy** [comment]
@goodfeli You can find the executable string below:

```
import warnings
warnings.filterwarnings('ignore')

import os
from cleverhans.attacks import CarliniWagnerL2
from io import BytesIO
import numpy as np
import pandas as pd
from PIL import Image
from scipy.misc import imsave,imread
import tensorflow as tf
from tensorflow.contrib.slim.nets import inception  

slim = tf.contrib.slim
tensorflow_master = ""
checkpoint_incv3       = "Pretrained/inception_v3.ckpt"
input_dir         = "dataset/images/"
categories = pd.read_csv("dataset/categories.csv")
image_classes = pd.read_csv("dataset/images.csv")

image_width       = 299
image_height      = 299
batch_size        = 32
batch_shape = [batch_size, image_height, image_width, 3]
num_classes = 1001


def load_images(input_dir, batch_shape):
    images = np.zeros(batch_shape)
    filenames = []
    labels =[]
    idx = 0
    batch_size = batch_shape[0]
    for filepath in sorted(tf.gfile.Glob(os.path.join(input_dir, '*.png'))):
        with tf.gfile.Open(filepath, "rb") as f:
            images[idx, :, :, :] = imread(f, mode='RGB').astype(np.float32)*2.0/255.0 - 1.0
        filenames.append(os.path.basename(filepath))
        labels.append(image_classes.merge(pd.DataFrame({"ImageId":[os.path.basename(filepath)[:-4]]}), on="ImageId")["TrueLabel"][0])
        idx += 1
        if idx == batch_size:
            yield filenames, images, labels
            filenames = []
            labels =[]
            images = np.zeros(batch_shape)
            idx = 0
    if idx > 0:
        yield filenames, images, labels

class InceptionModel(object):
    def __init__(self, num_classes):
        self.num_classes = num_classes
        self.built = False

    def __call__(self, x_input):
        """Constructs model and return probabilities for given input."""
        reuse = True if self.built else None
        with slim.arg_scope(inception.inception_v3_arg_scope()):
            _, end_points = inception.inception_v3(
                            x_input, num_classes=self.num_classes, is_training=False,
                            reuse=reuse)
        self.built = True
        output = end_points['Predictions']
        probs = output.op.inputs[0]
        return probs

image_iterator = load_images(input_dir, batch_shape)

with tf.Session() as sess:
    filenames, images, labels= next(image_iterator)

    x_input = tf.placeholder(tf.float32, shape=batch_shape)

    cw_params = {'binary_search_steps': 3, 
                 'abort_early' : True, 
                 'max_iterations': 250, 
                 'learning_rate': 0.001, 
                 'batch_size': batch_size, 
                 'initial_const': 100, 
                 'nb_classes': num_classes, 
                 'confidence': 0., 
                 'clip_min': -1.0, 
                 'clip_max': 1.0 
                 } 

    model = InceptionModel(num_classes)
    clw2=CarliniWagnerL2(model) 
    x_adv = clw2.generate(x_input)

    sess.run(tf.global_variables_initializer())
    saver = tf.train.Saver(slim.get_model_variables())
    nontargeted_images = sess.run(x_adv, feed_dict={x_input: images})
```

**Comment by B-Yassine** [comment]
Try to set the confidence to for example 70 and the learning rate to 0.9
for example, so normally you would have a very noisy image, then try to
visualize it, just to be sure that the attack is working or not

**Comment by goodfeli** [comment]
Before I run it on my end, can you add:

    abs_diff = np.abs(nontargeted_images - images)
    print(abs_diff.min(), abs_diff.mean(), abs_diff.max())

and let me know what it prints?

**Comment by blgnksy** [comment]
I run the code and get the results by using @goodfeli shared. Now output is less bright than the original but can classified to correct label. The Euclidean distance is nearly 196 which I think is very high.

```
(0.0, 0.22711349776335532, 1.0)
```
This is the result for my own params:

![outputs](https://user-images.githubusercontent.com/11997073/49919998-a70f4200-feb9-11e8-9e9d-855945879c20.png)

Then I set confidence to 70 and the learning rate to 0.9 but the results are exactly the same as the previous one. @B-Yassine

**Comment by B-Yassine** [comment]
it looks like the attack is not working, i will check your code, however, if you need it now you can check my simple coed of the C&W attack using cleverhans on the Keras Inceptionv3 pre-trained model [Here] (https://github.com/B-Yassine/Carlini-and-Wagner_InceptionV3_Imagenet)

**Comment by blgnksy** [comment]
Thank you @B-Yassine. I changed the parameters in your code example. And now the attack is working as expected. But still have problem working with Tensorflow. It is also weird that Keras backend is  Tensorflow and still works.

**Comment by B-Yassine** [comment]
Happy that it works, try maybe to upgrade your tensorflow version: `pip install --upgrade tensorflow-gpu==1.11.0`  just to avoid the last version "1.12.0" as i had a lot of errors with it that is why i downgraded to the previous version. Best,

**Comment by npapernot** [comment]
Looks like the issue was resolved but feel free to reopen with a reproducible script if you are still facing an issue.
